### ISSUE TITLE
FPs with relative links

### ISSUE THREAD

**Issue by XhmikosR** [JustinBeckwith/linkinator#135]
Not sure when this started happening exactly, but it seems that we get hundreds of errors with the latest version: https://github.com/twbs/bootstrap/runs/381540733#step:10:25

I know for sure this didn't happen some time ago because I have this branch around for some time now, but I can't recall when the issue started to happen.

**Comment by JustinBeckwith** [comment]
Thanks for the report @XhmikosR! Is this still an issue?  Could I trouble you with steps for a repro?

**Comment by XhmikosR** [comment]
Unfortunately, yeah :/

https://github.com/twbs/bootstrap/runs/435623930?check_suite_focus=true#step:10:8

**Comment by XhmikosR** [comment]
```
git clone https://github.com/twbs/bootstrap.git -b master-xmr-linkinator-prod
npm i
npm run docs-linkinator-prod
```

**Comment by XhmikosR** [comment]
Actually, even simpler:

```
linkinator https://getbootstrap.com --recurse --silent --skip "^(https://developer.microsoft.com)|(https://github.com/twbs/bootstrap/releases/)"
```

**Comment by XhmikosR** [comment]
Here's a scan I just ran with linkinator 2.0.3 https://github.com/twbs/bootstrap/runs/476650945#step:10:662

**Comment by JustinBeckwith** [comment]
So this is weird.  I just did a local run, on my machine, and I get:

```

> bootstrap@4.3.1 docs-linkinator-prod /Users/beckwith/Code/bootstrap
> linkinator https://getbootstrap.com --recurse --silent --skip "^(https://developer.microsoft.com)|(https://github.com/twbs/bootstrap/releases/)"

[403] https://codepen.io/Johann-S/pen/djJYPb
[403] http://accessibleculture.org/articles/2010/05/in-page-links/
[0] https://kkbruce.tw/bs3/
[0] http://salvattore.com/
[404] https://getbootstrap.com/2.3.0/assets/bootstrap.zip
[404] https://getbootstrap.com/2.2.1/assets/bootstrap.zip
[503] http://www.wordsbyf.at/2012/03/08/why-less/
[404] https://getbootstrap.com/2.1.1/assets/bootstrap.zip
[404] https://getbootstrap.com/2.1.0/assets/bootstrap.zip
[404] https://getbootstrap.com/2.0.4/assets/bootstrap.zip
[404] https://getbootstrap.com/2.0.3/assets/bootstrap.zip
[404] https://getbootstrap.com/2.3.1/assets/bootstrap.zip
[404] https://getbootstrap.com/2.2.0/assets/bootstrap.zip
[404] https://getbootstrap.com/2.2.2/assets/bootstrap.zip
[0] https://glyph.smarticons.co/
[404] https://dev.twitter.com/blog/bootstrap-twitter
[404] https://wpdev.uservoice.com/forums/257854-microsoft-edge-developer/suggestions/6508895
[404] https://wpdev.uservoice.com/forums/257854-microsoft-edge-developer/suggestions/12299532
[404] https://wpdev.uservoice.com/forums/257854-microsoft-edge-developer/suggestions/15939898
[404] https://wpdev.uservoice.com/forums/257854-microsoft-edge-developer/suggestions/15944476
[0] https://useiconic.com/open/
[0] https://useiconic.com/open
[404] https://wpdev.uservoice.com/forums/257854-microsoft-edge-developer/suggestions/6263621
```

The majority of those are in fact broken links.  The other few I am guessing are either client side redirects, or rejections based on user agent (which there is a fix floating around for).  

One thing we have observed, is that high CPU consumption can lead to an increase in request timeouts, which results in those response codes of `0` you're seeing.  My suspicion right now is that this is related to timeouts and performance.  Could I trouble you to try setting `--concurrency 10`?  Also make extra sure (I see you are, but still) that you're on `2.0.3`, because we recently fixed some nasty performance / heap bugs.

**Comment by XhmikosR** [comment]
@JustinBeckwith I also had the same results as you, which made this thing more confusing: when I run this from my machine I don't get so many failures. That being said, this issue did not happen in the past. Also, apart from the links you showed above, all other failed ones are relative, that's why I thought something is broken in this regard.

Here is the output with `--concurrency 10`, with fewer broken links AFAICT: https://github.com/twbs/bootstrap/runs/477598042#step:10:103

**Comment by XhmikosR** [comment]
Also, it seems I'm getting different number of links locally vs CI

* Locally:

    ERROR: Detected 40 broken links. Scanned **2511** links in 110.259 seconds.

* CI:

    ERROR: Detected 91 broken links. Scanned **2276** links in 261.197 seconds.

(Ignore the number of failures locally it's due to my network rules)

I'm on Windows for what is worth.

**Comment by XhmikosR** [comment]
Another issue I noticed when running locally is that failed URLs are only reported once. I have pi-hole set up, so since Google analytics is blocked shouldn't this be shown as broken for all pages? I should probably open a new issue about this if you think it makes sense

**Comment by JustinBeckwith** [comment]
Ok - so I think I know what's happening here, because we experienced something similar.  When the process/machine gets overwhelmed by outgoing connections, connections just start to get dropped, and they're returned as failures.  By default, we use a concurrency of 100, which works fine on my laptop, or on most servers.  However - in virtualized environments, like GitHub Actions I've noticed, the machines can handle far less.  We've bumped our concurrency down to 10 as a way to cope in these environments. 

Could I trouble you to give that a try?

**Comment by XhmikosR** [comment]
@JustinBeckwith yeah I tried that after that Twitter DM we had but still gets false positives, probably less, though 🙁

https://github.com/twbs/bootstrap/runs/641778881?check_suite_focus=true#step:8:1
